commit a2cad19f005be113b6b97aa60ca5b989ea1b0c4d
Author: Miroslav Rezanina <mrezanin@redhat.com>
Date:   Wed Jul 10 14:21:29 2013 +0200

    ga_install_service(): nest error paths more idiomatically
    
    RH-Author: Laszlo Ersek <lersek@redhat.com>
    Message-id: <1372954619-30116-5-git-send-email-lersek@redhat.com>
    O-Subject: [EMBGD RHEL-6.4.z qemu-kvm PATCH 4/5] ga_install_service(): nest error paths more idiomatically
    Bugzilla: 980758
    RH-Acked-by: Eric Blake <eblake@redhat.com>
    RH-Acked-by: Petr Matousek <pmatouse@redhat.com>
    RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
    
    Acked-by: Michael Roth <mdroth@linux.vnet.ibm.com>
    Reviewed-by: Eric Blake <eblake@redhat.com>
    Signed-off-by: Laszlo Ersek <lersek@redhat.com>

diff --git a/qga/service-win32.c b/qga/service-win32.c
index 255eeb3..94089a1 100644
--- a/qga/service-win32.c
+++ b/qga/service-win32.c
@@ -37,10 +37,12 @@ static int printf_win_error(const char *text)
 
 int ga_install_service(const char *path, const char *logfile)
 {
+    int ret = EXIT_FAILURE;
     SC_HANDLE manager;
     SC_HANDLE service;
     TCHAR module_fname[MAX_PATH];
     GString *cmdline;
+    SERVICE_DESCRIPTION desc = { (char *)QGA_SERVICE_DESCRIPTION };
 
     if (GetModuleFileName(NULL, module_fname, MAX_PATH) == 0) {
         printf_win_error("No full path to service's executable");
@@ -62,28 +64,28 @@ int ga_install_service(const char *path, const char *logfile)
     manager = OpenSCManager(NULL, NULL, SC_MANAGER_ALL_ACCESS);
     if (manager == NULL) {
         printf_win_error("No handle to service control manager");
-        g_string_free(cmdline, TRUE);
-        return EXIT_FAILURE;
+        goto out_strings;
     }
 
     service = CreateService(manager, QGA_SERVICE_NAME, QGA_SERVICE_DISPLAY_NAME,
         SERVICE_ALL_ACCESS, SERVICE_WIN32_OWN_PROCESS, SERVICE_AUTO_START,
         SERVICE_ERROR_NORMAL, cmdline->str, NULL, NULL, NULL, NULL, NULL);
-
-    if (service) {
-        SERVICE_DESCRIPTION desc = { (char *)QGA_SERVICE_DESCRIPTION };
-        ChangeServiceConfig2(service, SERVICE_CONFIG_DESCRIPTION, &desc);
-
-        fprintf(stderr, "Service was installed successfully.\n");
-    } else {
+    if (service == NULL) {
         printf_win_error("Failed to install service");
+        goto out_manager;
     }
 
+    ChangeServiceConfig2(service, SERVICE_CONFIG_DESCRIPTION, &desc);
+    fprintf(stderr, "Service was installed successfully.\n");
+    ret = EXIT_SUCCESS;
     CloseServiceHandle(service);
+
+out_manager:
     CloseServiceHandle(manager);
 
+out_strings:
     g_string_free(cmdline, TRUE);
-    return (service == NULL);
+    return ret;
 }
 
 int ga_uninstall_service(void)
